src/Microsoft.ML.AutoML.Interactive/NotebookMonitor.cs

```diff
@@ -0,0 +1,101 @@
+{
+    public class NotebookMonitor : IMonitor
+    {
+        private DisplayedValue _valueToUpdate;
+        private DateTime _lastUpdate = DateTime.MinValue;
+
+        public TrialResult BestTrial { get; set; }
+        public TrialResult MostRecentTrial { get; set; }
+        public TrialSettings ActiveTrial { get; set; }
+        public List<TrialResult> CompletedTrials { get; set; }
+        public DataFrame DataFrame { get; set; }
+
+        public NotebookMonitor()
+        {
+            this.CompletedTrials = new List<TrialResult>();
+            this.DataFrame = new DataFrame(new PrimitiveDataFrameColumn<int>("Trial"), new PrimitiveDataFrameColumn<float>("Metric"), new StringDataFrameColumn("Trainer"), new StringDataFrameColumn("Parameters"));
+        }
+
+        public void ReportBestTrial(TrialResult result)
+        {
+            this.BestTrial = result;
+            Update();
+        }
+
+        public void ReportCompletedTrial(TrialResult result)
+        {
+            this.MostRecentTrial = result;
+            this.CompletedTrials.Add(result);
+
+            var activeRunParam = JsonSerializer.Serialize(result.TrialSettings.Parameter, new JsonSerializerOptions() { WriteIndented = false, });
+
+            this.DataFrame.Append(new List<KeyValuePair<string, object>>()
+            {
+                new KeyValuePair<string, object>("Trial",result.TrialSettings.TrialId),
+                new KeyValuePair<string, object>("Metric", result.Metric),
+                new KeyValuePair<string, object>("Trainer",result.TrialSettings.Pipeline.ToString().Replace("Unknown=>","")),
+                new KeyValuePair<string, object>("Parameters",activeRunParam),
+            }, true);
+            Update();
+        }
+
+        public void ReportFailTrial(TrialResult result)
+        {
+            // TODO figure out what to do with failed trials.
+            Update();
+        }
+
+        public void ReportRunningTrial(TrialSettings setting)
+        {
+            this.ActiveTrial = setting;
+            Update();
+        }
+
+        private bool _updatePending = false;
+        public void Update()
+        {
+            Task.Run(async () =>
+            {
+                if (_updatePending == true)
+                {
+                    // Keep waiting
+                }
+                else
+                {
+                    int timeRemaining = 5000 - (int)(DateTime.Now.Millisecond - this._lastUpdate.Millisecond);
+                    _updatePending = true;
+                    if (timeRemaining > 0)
+                    {
+                        await Task.Delay(timeRemaining);
+                    }
+                    if (this._valueToUpdate != null)
+                    {
+                        _updatePending = false;
+                        this._lastUpdate = DateTime.Now;
+                        this._valueToUpdate.Update(this);
+                    }
+                }
+            });
+
+        }
```

Comment:
Is there a better way to throttle in .NET? I essentially just want to make sure this runs no more than 1 time in 5 seconds.

Reply:
The if check was mainly to avoid the exception if it's negative.

https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.delay?view=net-6.0#system-threading-tasks-task-delay(system-int32)

> Exceptions
> [ArgumentOutOfRangeException](https://docs.microsoft.com/en-us/dotnet/api/system.argumentoutofrangeexception?view=net-6.0)
 The millisecondsDelay argument is less than -1.

Reply:
CC @stephentoub if you has better recommendation to do that

Reply:
Moved to: https://github.com/dotnet/machinelearning/pull/6243/files#diff-38e7b5c79365be27915c2761266d21f78921c344992732ed2f0b2a4360f4ddc2R34